libbeat/processors/extract_array/extract_array.go

```diff
@@ -0,0 +1,178 @@
+		colIdx, ok := common.TryToInt(column)
+		if !ok || colIdx < 0 {
+			return nil, fmt.Errorf("bad extract_array mapping for field %s: %+v is not a positive integer", field, column)
+		}
+		p.mappings = append(p.mappings, fieldMapping{from: colIdx, to: field})
+	}
+	sort.Slice(p.mappings, func(i, j int) bool {
+		return p.mappings[i].from < p.mappings[j].from
+	})
+	return p, nil
+}
+
+func isEmpty(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.String:
+		return v.Len() == 0
+	case reflect.Slice, reflect.Map:
+		return v.IsNil() || v.Len() == 0
+	case reflect.Interface:
+		return v.IsNil() || isEmpty(v.Elem())
+	}
+	return false
+}
+
+func (f *extractArrayProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	iValue, err := event.GetValue(f.config.Field)
+	if err != nil {
+		if f.config.IgnoreMissing && errors.Cause(err) == common.ErrKeyNotFound {
+			return event, nil
+		}
+		return event, errors.Wrapf(err, "could not fetch value for field %s", f.config.Field)
+	}
+
+	array := reflect.ValueOf(iValue)
+	if t := array.Type(); t.Kind() != reflect.Slice {
+		if !f.config.FailOnError {
+			return event, nil
+		}
+		return event, errors.Wrapf(err, "unsupported type for field %s: got: %s needed: array", f.config.Field, t.String())
+	}
+
+	saved := *event
+	if f.config.FailOnError {
+		saved.Fields = event.Fields.Clone()
+		saved.Meta = event.Meta.Clone()
+	}
+
+	n := array.Len()
+	for _, mapping := range f.mappings {
+		if mapping.from >= n {
+			if !f.config.FailOnError {
+				continue
+			}
+			return &saved, errors.Errorf("index %d exceeds length of %d when processing mapping for field %s", mapping.from, n, mapping.to)
+		}
+		cell := array.Index(mapping.from)
+		// checking for CanInterface() here is done to prevent .Interface() from
+		// panicking, but it can only happen when value points to a private
+		// field inside a struct.
+		if !cell.CanInterface() || (f.config.OmitEmpty && isEmpty(cell)) {
+			continue
+		}
+		if !f.config.OverwriteKeys {
+			if _, err = event.GetValue(mapping.to); err == nil {
+				if !f.config.FailOnError {
+					continue
+				}
+				return &saved, errors.Errorf("target field %s already has a value. Set the overwrite_keys flag or drop/rename the field first", mapping.to)
+			}
+		}
+		if _, err = event.PutValue(mapping.to, clone(cell.Interface())); err != nil {
+			if !f.config.FailOnError {
+				continue
+			}
+			return &saved, errors.Wrapf(err, "failed setting field %s", mapping.to)
+		}
+	}
+	return event, nil
+}
+
+func (f *extractArrayProcessor) String() (r string) {
+	return fmt.Sprintf("extract_array={field=%s, mappings=%v}", f.config.Field, f.mappings)
+}
+
+func clone(value interface{}) interface{} {
+	// TODO: This is dangerous but done by most processors.
+	//       Otherwise need to reflect value and deep copy lists / map types.
+	switch v := value.(type) {
+	case common.MapStr:
+		return v.Clone()
+	case map[string]interface{}:
+		return common.MapStr(v).Clone()
+	}
+	return value
+}
```

Comment:
I presume we use clone, so modifications don't affect the original data within the array? In this case also look out for arrays. Some processors like `add_tags` can modify arrays.